### ISSUE TITLE
JuMP.StopTheSolver only works for addlazycallback?

### ISSUE THREAD

**Issue by haoxiangyang89** [JuliaOpt/JuMP.jl#1144]
I am running a branch and cut algorithm and I would like to have the ability to stop the callback at a MIPNode. addlazycallback will only be triggered at a MIPSol so I have an additional cutcallback which can be triggered at a MIPNode, but 
`return JuMP.StopTheSolver`
within the cutcallback will not stop the solving process and return the :UserLimit as expected.

Here is the example from JuMP guide:
```julia
using JuMP
using Gurobi

# We will use Gurobi, which requires that we manually set the attribute
# PreCrush to 1 if we have user cuts. We will also disable PreSolve, Cuts,
# and Heuristics so only our cut will be used
m = Model(solver=GurobiSolver(PreCrush=1, Cuts=0, Presolve=0, Heuristics=0.0))

# Define our variables to be inside a box, and integer
@variable(m, 0 <= x <= 2, Int)
@variable(m, 0 <= y <= 2, Int)

# Optimal solution is trying to go towards top-right corner (2.0, 2.0)
@objective(m, Max, x + 2y)

# We have one constraint that cuts off the top right corner
@constraint(m, y + x <= 3.5)

# Optimal solution of relaxed problem will be (1.5, 2.0)
# We can add a user cut that will cut of this fractional solution.

# We now define our callback function that takes one argument,
# the callback handle. Note that we can access m, x, and y because
# this function is defined inside the same scope
function mycutgenerator(cb)
    x_val = getvalue(x)
    y_val = getvalue(y)
    println("In callback function, x=$x_val, y=$y_val")

    # Allow for some impreciseness in the solution
    TOL = 1e-6

    # Check top right
    if y_val + x_val > 3 + TOL
        # Cut off this solution
        println("Fractional solution was in top right, cut it off")
        # Use the original variables
        return JuMP.StopTheSolver
    end
end  # End of callback function

# Tell JuMP/Gurobi to use our callback function
addcutcallback(m, mycutgenerator)

# Solve the problem
solve(m)

# Print our final solution
println("Final solution: [ $(getvalue(x)), $(getvalue(y)) ]")
```

Here I substitute the original usercut sentence by " return JuMP.StopTheSolver". Instead of stop at the solution x = 1.5, y = 2.0, the process continues until it reaches the final solution x = 1, y = 2. I am not sure if this is affected by the current reconstruction or it has been this way since the beginning.

Thanks!

**Comment by joehuchette** [comment]
Are you sure that the callback function is actually being called? Gurobi and CPLEX don't offer guarantees when/if this will happen.

**Comment by haoxiangyang89** [comment]
I actually fixed this problem on my local end. Cplex/Gurobi is actually called. The error is due to this: In callback.jl, we have a treatment when the StopTheSolver is returned (line 76 - 83):

```
for cb in cbs
            if state == :MIPSol || cb.fractional
                ret = cb.f(d)
                if ret === StopTheSolver
                    return :Exit
                end
            end
        end
```

This is not present in other type of callbacks (heuristic/info/cut). Simply adding the quoted part to each of them would solve the problem. I am not sure about the current state of development of the JuMP so let me know if you want to me to create a pull request or you want to change it on your end. 

One interesting point though. Since the code above is essentially returning :Exit when StopTheSolver is encountered, I tried directly `return :Exit` when I need to stop the solver. There it does not work and I am not sure why.

**Comment by mlubin** [comment]
@haoxiangyang89, seems like a reasonable fix. We'll consider pull requests to the `release-0.18` branch.

**Comment by joehuchette** [comment]
Agreed, it looks like the ``ret === StopTheSolver`` check is missing from ``cutcallback``, ``heurcallback``, and ``infocallback``, so a PR (against ``release-0.18``) would be appreciated.

**Comment by mlubin** [comment]
This is fixed in the `release-0.18` branch.
